crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk_extension.move

```diff
@@ -0,0 +1,252 @@
+/// prevent asset locking.
+/// - `kiosk_extension` is a friend module to `kiosk` and has access to its
+/// internal functions (such as `place_internal` and `lock_internal` to
+/// implement custom authorization scheme for `place` and `lock` respectively).
+module sui::kiosk_extension {
+    use sui::bag::{Self, Bag};
+    use sui::dynamic_field as df;
+    use sui::tx_context::TxContext;
+    use sui::transfer_policy::TransferPolicy;
+    use sui::kiosk::{Self, Kiosk, KioskOwnerCap};
+
+    /// Trying to add an extension while not being the owner of the Kiosk.
+    const ENotOwner: u64 = 0;
+    /// Extension is trying to access a permissioned action while being disabled.
+    const EExtensionDisabled: u64 = 1;
+    /// Extension is trying to access a permissioned action while not having
+    /// the required permission.
+    const EExtensionNotAllowed: u64 = 2;
+    /// Extension is not installed in the Kiosk.
+    const EExtensionNotInstalled: u64 = 3;
+
+    /// The Extension struct contains the data used by the extension and the
+    /// configuration for this extension. Stored under the `ExtensionKey`
+    /// dynamic field.
+    struct Extension has store {
+        /// Storage for the extension, an isolated Bag. By putting the extension
+        /// into a single dynamic field, we reduce the amount of fields on the
+        /// top level (eg items / listings) while giving extension developers
+        /// the ability to store any data they want.
+        storage: Bag,
+        /// Bitmap of permissions that the extension has (can be revoked any
+        /// moment). It's all or nothing policy - either the extension has the
+        /// required permissions or no permissions at all.
+        ///
+        /// 1st bit - `place` - allows to place items for sale
+        /// 2nd bit - `lock` - allows to lock items
+        ///
+        /// For example:
+        /// - `11` - allows to place items and lock them.
+        /// - `01` - allows to place items, but not lock them.
+        /// - `00` - no permissions.
+        permissions: u32,
+        /// Whether the extension can call protected actions. By default, all
+        /// extensions are enabled (on `add` call), however the Kiosk
+        /// owner can disable them at any time.
+        ///
+        /// Disabling the extension does not limit its access to the storage.
+        is_enabled: bool,
+    }
+
+    /// The `ExtensionKey` is a typed dynamic field key used to store the
+    /// extension configuration and data. `Ext` is a phantom type that is used
+    /// to identify the extension witness.
+    struct ExtensionKey<phantom Ext> has store, copy, drop {}
+
+    // === Management ===
+
+    /// Add an extension to the Kiosk. Can only be performed by the owner.
+    ///
+    /// Unlike the original implementation, this one uses the `public`
+    /// visibility modifier to allow "owned kiosk" extension. Ideally we don't
```

Comment:
nit: This is useful context for the review, but I would remove reference to an "original implementation" before landing this, because it's going to be confusing for users of the API who won't know or care about the fact that there was a previous version of this code.

Reply:
Thanks! Yeah, this one was more like a note to self or to reviewer.